scipy/sparse/linalg/_isolve/tests/test_lsqr.py

```diff
@@ -20,12 +20,12 @@
 
 b = normal(size=n)
 
-tol = 1e-10
+tol = 2e-10
```

Comment:
For linux+MKL+AVX512, this needs to be:
```suggestion
tol = 1e-8
```
cf.
```
E   AssertionError: 
E   Not equal to tolerance rtol=2e-10, atol=2e-10
E   
E   Mismatched elements: 2 / 35 (5.71%)
E   Max absolute difference: 3.43928219e-10
E   Max relative difference: 1.12942507e-08
```

Reply:
This proposed bump is quite large. I thought about special-casing MKL, but that's too much of a pain (need `threadpoolctl` to even detect if we're using MKL). I think splitting into `atol=4e-10` and `rtol=1.5e-8` makes sense here.

Reply:
Actually, the failure is weird. `tol` is the _requested_ tolerance to `lsqr`. Having the returned tolerance be 2 orders of magnitude of is odd - it may be that the tolerance in the test itself is going haywire.

Reply:
@h-vetinari have you tested your proposed tolerance bump? If not, then I'd expect it to not necessarily resolve the problem.

Reply:
> This proposed bump is quite large. [...] I think splitting into `atol=4e-10` and `rtol=1.5e-8` makes sense here.

Seems you're talking about something else in this comment? I was talking about `1e-10` -> `1e-8` which I didn't consider _that_ large.

> Actually, the failure is weird. `tol` is the _requested_ tolerance to `lsqr`. Having the returned tolerance be 2 orders of magnitude of is odd - it may be that the tolerance in the test itself is going haywire.

Yeah, it's _both_ the tolerance requested from the routine, as well as the tolerance required to pass - i.e. no degradation allowed at all. While this is possible for some `while err > tol: iterate()` scenarios, it seems even the lsqr docstring warns against this:
>  - **atol, btol: _float, optional_**
    Stopping tolerances. If both are 1.0e-9 (say), the final residual norm should be accurate to about 9 digits. (The final x will usually have fewer correct digits, depending on cond(A) and the size of damp.)

In other words, the test seems is badly designed. Indeed, I checked the condition of the matrix `G` being used in that test, and it's ~280, which roughly matches the loss from `1e-10` to `2e-8`. I guess other platforms/arches/blas flavours are automatically below `1e-10` (despite the condition of `G`; due to the intrinsic double eps being small enough). The fact that it only comes out with AVX512 could however be an indicator that certain functions (which I believe come with different accuracy guarantees like 1ULP, 5ULP, etc.) might be a bit on the loose side.

> @h-vetinari have you tested your proposed tolerance bump? If not, then I'd expect it to not necessarily resolve the problem.

I added a commit to check in https://github.com/conda-forge/scipy-feedstock/pull/198 before doing the investigation above. Guess it might still work due to the the same kind of "background accuracy" effect.

Reply:
Update: indeed, bumping to `2e-8` makes the test pass also on linux+MKL+AVX512. I'd consider this a stop gap, given that the test seems badly set up and probably needs some more care & attention.

Reply:
thanks! happy with a stopgap, we have many of those:)

Reply:
Argh... bumping the tolerance to `2e-8` makes _other_ blas variants gain a failure that in turn depends on the tolerance being passed to `lsqr` being small enough:
```
_____________________________ test_initialization ______________________________
../]...]/lib/python3.8/site-packages/scipy/sparse/linalg/_isolve/tests/test_lsqr.py:115: in test_initialization
    assert_allclose(x_ref[0], x[0])
E   AssertionError: 
E   Not equal to tolerance rtol=1e-07, atol=0
E   
E   Mismatched elements: 1 / 35 (2.86%)
E   Max absolute difference: 3.65397923e-09
E   Max relative difference: 1.20581449e-07
```
This happens not least because `test_initialization` is using `assert_allclose` with its default tolerances
```python
    # Test warm-start with single iteration
    x0 = lsqr(G, b, show=show, atol=tol, btol=tol, iter_lim=1)[0]
    x = lsqr(G, b, show=show, atol=tol, btol=tol, iter_lim=maxit, x0=x0)
    assert_allclose(x_ref[0], x[0])
    assert_array_equal(b_copy, b)
```
which then creates a silent coupling to the tolerance set at the top.

Reply:
hmm, okay - I just pushed a commit. let me try to split it out then

Reply:
Tried with:
```
# tolerance for atol/btol keywords of lsqr()
tol = 2e-10
# tolerances for testing the results of the lsqr() call with assert_allclose
# These tolerances are a bit fragile - see discussion in gh-15301.
atol_test = 4e-10
rtol_test = 2e-8
```

Reply:
Okay that failed quickly:
```_____________________________ test_initialization ______________________________
[gw0] linux -- Python 3.9.9 /opt/hostedtoolcache/Python/3.9.9/x64/bin/python
../../installdir/lib/python3.9/site-packages/scipy/sparse/linalg/_isolve/tests/test_lsqr.py:118: in test_initialization
    assert_allclose(x_ref[0], x[0])
E   AssertionError: 
E   Not equal to tolerance rtol=1e-07, atol=0
E   
E   Mismatched elements: 1 / 35 (2.86%)
E   Max absolute difference: 3.65397923e-09
E   Max relative difference: 1.20581449e-07
E    x: array([ 1.24264 ,  0.01353 , -0.744982,  1.472967, -0.027933, -1.758359,
E           1.554676,  1.404636,  0.057257,  0.040355,  0.760972,  0.902545,
E          -2.890788, -1.010879,  0.508212,  0.895522, -0.423499, -1.371109,...
E    y: array([ 1.24264 ,  0.01353 , -0.744982,  1.472967, -0.027933, -1.758359,
E           1.554676,  1.404636,  0.057257,  0.040355,  0.760972,  0.902545,
E          -2.890788, -1.010879,  0.508212,  0.895522, -0.423499, -1.371109,...
        b_copy     = array([ 1.53877394,  0.30966399, -0.44884868,  1.76910057,  0.26820049,
       -1.46222506,  1.85081017,  1.70076951, ...318443, -0.2003623 ,  0.54567   ,  0.26228662,
        0.21907021,  0.93235293,  1.74305243, -0.12470368,  0.95403167])
        x          = (array([ 1.24264013,  0.01353017, -0.7449825 ,  1.47296675, -0.02793333,
       -1.75835888,  1.55467635,  1.40463569,...,  1.44691861, -0.4208375 ,  0.65789785]), 1, 3, 9.900828998994749e-12, 9.900828998994749e-12, 215.89522744093603, ...)
        x0         = array([ 2.34680389e-03,  1.44041209e-03,  1.18795124e-03,  2.01387746e-03,
        6.74005409e-04,  1.20893788e-03, -3...2512e-04,  1.48223786e-03,  2.23351686e-03,  1.08262178e-03,
        1.66191324e-03,  2.57137204e-03,  1.35363149e-03])
        x_ref      = (array([ 1.24264012,  0.01353017, -0.7449825 ,  1.47296675, -0.02793333,
       -1.75835888,  1.55467635,  1.40463569,...2,  1.44691861, -0.4208375 ,  0.65789785]), 1, 3, 2.423210229620179e-06, 2.423210229620179e-06, 215.8952274409364, ...)

FAILED ../../installdir/lib/python3.9/site-packages/scipy/sparse/linalg/_isolve/tests/test_lsqr.py::test_initialization
```

Reply:
See my comment further up. `test_initialization` implicitly depends on having a small global `tol` because it uses default tolerances in `assert_allclose`. I think "unbumping" that, while keeping the separate tolerances you just introduced should do the trick.

Reply:
Actually `tol=2e-10` (+the `_test` tolerances) seems to work for the conda-forge infra! :)

I also see green CI here (disregarding the timeouts), so I'm not sure where you saw this failure?

Reply:
Hmm, in a notification - maybe it was for an older commit. It indeed looks good now.